src/v/resource_mgmt/cpu_profiler.cc

```diff
@@ -0,0 +1,153 @@
+#include "ssx/future-util.h"
+#include "ssx/sformat.h"
+
+#include <seastar/core/future.hh>
+#include <seastar/core/internal/cpu_profiler.hh>
+#include <seastar/core/reactor.hh>
+#include <seastar/util/later.hh>
+
+#include <iterator>
+
+namespace resources {
+
+cpu_profiler::cpu_profiler(
+  config::binding<bool>&& enabled,
+  config::binding<std::chrono::milliseconds>&& sample_rate)
+  : _query_timer([this] { poll_samples(); })
+  , _enabled(std::move(enabled))
+  , _sample_rate(std::move(sample_rate)) {
+    _enabled.watch([this] { on_enabled_change(); });
+    _sample_rate.watch([this] { on_sample_rate_change(); });
+}
+
+ss::future<> cpu_profiler::start() {
+    seastar::engine().set_cpu_profiler_period(_sample_rate());
+    seastar::engine().set_cpu_profiler_enabled(_enabled());
+
+    if (_enabled()) {
+        // Arm the timer to fire whenever the profiler collects
+        // the maximum number of traces it can retain.
+        _query_timer.arm_periodic(ss::max_number_of_traces * _sample_rate());
+    }
+
+    return ss::now();
+}
+
+ss::future<> cpu_profiler::stop() {
+    _query_timer.cancel();
+    co_await _gate.close();
+}
+
+void cpu_profiler::poll_samples() { do_poll_samples(); }
+
+ss::future<cpu_profiler::cpu_profiler_results>
+cpu_profiler::results(std::optional<ss::shard_id> shard_id) {
+    if (_gate.is_closed()) {
+        co_return cpu_profiler_results{};
+    }
+    auto holder = _gate.hold();
+
+    using result_t = std::pair<ss::shard_id, std::vector<cpu_profiler_result>>;
+    cpu_profiler_results results{};
+
+    if (shard_id) {
+        results.emplace_back(
+          shard_id.value(),
+          co_await container().invoke_on(
+            shard_id.value(), [](auto& s) { return s.shard_results(); }));
+    } else {
+        results = co_await container().map_reduce0(
+          [](auto& s) {
+              return result_t{ss::this_shard_id(), s.shard_results()};
+          },
+          cpu_profiler_results{},
+          [](cpu_profiler_results results, result_t shard_result) {
+              results.emplace_back(
+                shard_result.first, std::move(shard_result.second));
+              return results;
+          });
+    }
+
+    co_return results;
+}
+
+std::vector<cpu_profiler::cpu_profiler_result>
+cpu_profiler::shard_results() const {
+    absl::node_hash_map<ss::simple_backtrace, size_t> backtraces;
+    for (auto& results_buffer : _results_buffers) {
+        for (auto& result : results_buffer) {
+            auto [it, ins] = backtraces.try_emplace(result.user_backtrace, 1);
+            if (!ins) {
+                it->second++;
+            }
+        }
+    }
+
+    std::vector<cpu_profiler_result> results{};
+    results.reserve(backtraces.size());
+
+    for (auto& backtrace : backtraces) {
+        results.emplace_back(
+          ssx::sformat("{}", backtrace.first), backtrace.second);
+    }
+
+    return results;
+}
+
+void cpu_profiler::do_poll_samples() {
```

Comment:
So this service is periodically polling the underlying seastar profiling service and accumulating the results in a circular buffer?

Basically I'm wondering why there isn't just one place which accumulates samples.

Reply:
I believe the correct interface is to have a service that polls the seastar cpu profiler at set intervals. The other                     option would be having an interface in seastar that takes in some function that is executed

Reply:
I'm having seastar accumulate some samples first in order to reduce the amount of polling that needs to occur if a user sets the sample period to less then 1ms. At that point even if the application set a timer to poll seastar at 1ms or less it's unlikely that the reactor will be able to schedule the polls exact enough to prevent a good portion of the samples from being lost.

The other option would be having an interface in seastar that takes in some function that is executed during the interrupt or scheduled as a task to be executed via the reactor at some point after the interrupt occurred. Executing the user-provided function during the interrupt is a non-starter. So it'd probably have to be scheduled as a task to run at some point afterwards. At a period of 100ms per sample this would probably be fine. However, at some period less than 1ms this would schedule a lot of extra tasks that the reactor has to execute. So even with this interface some accumulation on the seastar side of things would probably be desirable.

Reply:
I was also trying to avoid having another configuration parameter in the seastar layer like `set_total_backtraces_to_retain(size_t)` to keep the interface/implementation minimal. So seastar just retains a fixed amount(128) to allow somewhat efficient sample periods less than or equal to 1ms.